keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,345 @@
+    - [Changes of existing components](#changes-of-existing-components)
+      - [Topology Manager changes](#topology-manager-changes)
+      - [Internal Container Lifecycle changes](#internal-container-lifecycle-changes)
+    - [Feature Gate and Kubelet Flags](#feature-gate-and-kubelet-flags)
+- [Graduation Criteria](#graduation-criteria)
+  - [Phase 1: Alpha (target v1.19)](#phase-1-alpha-target-v119)
+  - [Phase 2: Beta](#phase-2-beta)
+  - [GA (stable)](#ga-stable)
+- [Appendix](#appendix)
+  - [How DPDK works](#how-dpdk-works)
+  - [Requirements to guarantee performance of DPDK](#requirements-to-guarantee-performance-of-dpdk)
+  - [How Linux OOM system works](#how-linux-oom-system-works)
+  - [How Kubernetes utilizes Linux OOM system](#how-kubernetes-utilizes-linux-oom-system)
+<!-- /toc -->
+
+# Overview
+
+NUMA Awareness is well known for a solution to boost performance for diverse use cases including DPDK and Database on multi socket machine. Since Kubernetes 1.16, it started to support NUMA aware resource allocation through Topology Manager for NUMA sensitive containers.
+Topology Manager coordinates resource alignment by using hint interface that allows advertising the resource capability of NUMA node.
+In kubelet, CPU Manager and Device Manager advertise their resources with topology information as topology hint then Topology Manager calculates best set of topology hints for containers.
+But there's no feature for memory and hugepages. It leads kubelet to result NUMA Un-Aware memory and hugepages management for containers.
+It can cause inter-NUMA node communication for Memory access which leads to increase an I/O latency and decrease performance.
+
+The Memory Manager is proposed to provide a way to guarantee NUMA awareness for memory and hugepages usage to kubelet.
+
+# Motivation
+
+## Related Features
+
+- [Topology Manager][topology-manager] is a feature that collects topology hint from various hint providers to calculate NUMA affinity for a container. Topology Manager judge container whether admit or not under criteria of configured topology policy and calculated NUMA affinity.
+
+- [CPU Manager][cpu-manager] is a feature provides a solution for CPU Pinning based on cgroups cpuset subsystem, it also offer topology hint to Topology Manager.
+
+- [Device Manager][device-manager] is a one of features provides topology hint for Topology Manager. The main objectivity of Device Manager is allow vendors to adverise their resources and capability to kubelet.
+
+- [Hugepages][hugepages] is a feature allows container to consume pre-allocated hugepages with pod isolation of huge pages.
+
+- [Node Allocatable Feature][node-allocatable-feature] is a feature helps to reserve compute resources to prevent resource starvation. The kube-reserved and system-reserved is used to reserve resources for kubelet and system(OS, etc). At the moment(1.17), following resources are supported to reserve: cpu, memory, ephemeral storage.
```

Comment:
Do we support reserving these resources on a per-numa-node basis?  If not, are there plans to do so?